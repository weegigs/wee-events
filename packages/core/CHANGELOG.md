# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [0.18.1](https://github.com/weegigs/wee-events/compare/v0.18.0...v0.18.1) (2023-03-05)

### Bug Fixes

- correct revision returned by memory store when publishing multiple events ([00d4c2e](https://github.com/weegigs/wee-events/commit/00d4c2e4030a4b6053f29c214bc859d8bfdc3e95))

# 0.18.0 (2023-02-24)

### Bug Fixes

- correct depenencies ([97dfaab](https://github.com/weegigs/wee-events/commit/97dfaab70f4863b71b190f6be68bc75e72618803))
- entity revision should be the same as the last event processed, not evaluated ([2334381](https://github.com/weegigs/wee-events/commit/2334381d21dcab8c052210a6af74da5af96b62dc))
- revert nanoid to non-esmodule version ([3379e16](https://github.com/weegigs/wee-events/commit/3379e1631db6525a01f282fd3281947a6eedb02b))

### Features

- effects based service definition ([59fbe04](https://github.com/weegigs/wee-events/commit/59fbe0433839f220f5f9cc5aa43d0dc78e7c0c19))
