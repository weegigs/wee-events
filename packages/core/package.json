{
  "name": "@weegigs/events-core",
  "version": "0.16.1",
  "description": "Core Event System",
  "author": "Kevin O'Neill <kevin@weegigs.com.au>",
  "license": "MIT",
  "main": "./lib/index.js",
  "types": "./lib/index.d.ts",
  "packageManager": "yarn@3.2.1",
  "repository": {
    "type": "git",
    "url": "https://github.com/weegigs/wee-events.git"
  },
  "scripts": {
    "build": "yarn ws:build",
    "clean": "yarn ws:clean",
    "compile": "yarn ws:compile",
    "test": "yarn ws:test",
    "lint": "yarn ws:lint",
    "doctor": "yarn ws:doctor",
    "fix": "yarn run ws:fix",
    "prepack": "yarn run ws:prepack",
    "publish": "yarn run ws:publish"
  },
  "dependencies": {
    "lodash": "^4.17.21",
    "luxon": "^3.0.3",
    "nanoid": "^4.0.0",
    "ulid": "^2.3.0",
    "zod": "^3.19.1"
  },
  "devDependencies": {
    "@tsconfig/node16": "^1.0.3",
    "@types/jest": "^29.0.3",
    "@types/lodash": "^4.14.185",
    "@types/luxon": "^3.0.1",
    "@types/node": "^16.11.59",
    "jest": "^29.0.3",
    "ts-jest": "^29.0.1",
    "typescript": "^4.8.3"
  },
  "files": [
    "lib"
  ],
  "jest": {
    "preset": "ts-jest",
    "testEnvironment": "node",
    "collectCoverageFrom": [
      "**/*.{ts,tsx}",
      "!**/node_modules/**"
    ],
    "testMatch": [
      "**/*.spec.ts",
      "**/*.test.ts"
    ],
    "collectCoverage": false,
    "coverageReporters": [
      "html"
    ]
  },
  "prettier": {
    "printWidth": 120,
    "trailingComma": "es5"
  }
}
