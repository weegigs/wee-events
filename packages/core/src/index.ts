export * from "./errors";
export * from "./types";
export * from "./store";
export * from "./loader";
export * from "./service";
