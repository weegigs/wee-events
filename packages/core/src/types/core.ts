import * as z from "zod";

export type Primitive = string | number | boolean;
export type Value = Primitive | Payload | List;
export type EmptyPayload = { [member: string]: never };
export type Payload = { [member: string]: Value } | EmptyPayload;
export type List = Primitive[] | Payload[] | List[];

export namespace Payload {
  const primitive = z.union([z.string(), z.number(), z.boolean()]);
  const list: z.Schema<List> = z.lazy(() => z.union([z.array(primitive), z.array(payload), z.array(list)]));
  const value: z.Schema<Value> = z.lazy(() => z.union([primitive, list, payload]));
  const payload: z.Schema<Payload> = z.lazy(() => z.record(value));

  export const schema = payload;
  export const parse = (value: unknown): Payload => schema.parse(value);
}

export namespace EmptyPayload {
  export const schema = z.record(z.never());
}

export type AggregateId<Type extends string = string> = {
  type: Type;
  key: string;
};

export namespace AggregateId {
  export const schema = <Type extends string>(type: Type): z.Schema<AggregateId<Type>> =>
    z.object({
      type: z.literal(type),
      key: z.string().min(1),
    }) as unknown as z.Schema<AggregateId<Type>>;

  export const encode = (aggregate: AggregateId): string => `${aggregate.type}.${aggregate.key}`;
  export const decoder =
    (type: string) =>
    (aggregate: string): AggregateId => {
      const [_type, ...key] = aggregate.split(".");
      return schema(type).parse({ type: _type, key: key.join(".") });
    };

  export const parser =
    <Type extends string>(type: Type) =>
    (value: unknown) =>
      schema(type).parse(value);

  export function create<const T extends string>(type: T, key: string): AggregateId<T> {
    return { type, key } as const;
  }
}

export interface DomainEvent<Type extends string = string, Data extends Payload = Payload> {
  type: Type;
  data: Data;
}

export namespace DomainEvent {
  export const schema = <S extends z.Schema>(type: string, data: S) =>
    z.object({
      type: z.literal(type),
      data,
    });
}

export interface RecordedEvent<E extends DomainEvent = DomainEvent> {
  /**
   * The aggregate that events belongs to.
   */
  aggregate: AggregateId;

  /**
   * The revision for the aggregate.
   */
  revision: string;

  /**
   * Unique identifier representing this event. ULID format.
   */
  id: string;

  /**
   * Type of this event.
   */
  type: E["type"];

  /**
   * Representing when this event was created in the database system.
   */
  timestamp: string;

  /**
   * Data of this event.
   */
  data: E["data"];

  /**
   * Representing the metadata associated with this event.
   */
  metadata: RecordedEvent.Metadata;
}

export namespace RecordedEvent {
  export type Metadata = {
    causationId?: string;
    correlationId?: string;
  };
}

export type Revision = string;

export namespace Revision {
  export type Initial = "00000000000000000000000000";
  export const Initial = "00000000000000000000000000";

  export const isInitial = (revision: Revision): boolean => revision === Initial;

  export const schema = z.string().min(26).max(26);
}
