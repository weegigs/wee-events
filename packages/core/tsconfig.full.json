{
  "extends": "@tsconfig/node22/tsconfig.json",
  "compilerOptions": {
    "composite": true,
    "isolatedModules": true,
    "esModuleInterop": true,
    "removeComments": true,
    "sourceMap": true,
    "inlineSources": true,
    "declaration": true,
    "declarationMap": true,
    "experimentalDecorators": true,
    "emitDecoratorMetadata": true,
    "strict": true,
    "skipLibCheck": true,
    "exactOptionalPropertyTypes": true,
    "noImplicitOverride": true,
    "noImplicitReturns": true,
    "noUnusedLocals": true,
    "noUnusedParameters": true,
    "noEmit": false,
    "rootDir": "./src",
    "outDir": "./lib"
  },
  "include": ["scripts/**/*.ts"],
  "exclude": ["node_modules/**/*", ".build/**/*", ".vscode/**/*"],
  "include": [
    "src/**/*.ts",
    "../effects/entity-loader.ts"
  ],
  "exclude": [
    "./node_modules/**/*",
    "./lib/**/*",
    "./.vscode/**/*"
  ],
  "references": []
}

