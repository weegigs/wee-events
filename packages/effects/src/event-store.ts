import * as T from "@effect-ts/core/Effect";
import { Has, tag } from "@effect-ts/core/Has";
import { pipe } from "@effect-ts/core";

import * as wee from "@weegigs/events-core";

export const EventStore = tag<wee.EventStore>();
export type HasEventStore = Has<wee.EventStore>;

export interface EventPublisher {
  publish: wee.EventStore.Publisher;
}
export const EventPublisher = tag<EventPublisher>();

export interface EventLoader {
  load: wee.EventStore.Loader;
}
export const EventLoader = tag<EventLoader>();

export class LoaderError extends Error {
  constructor(public override readonly cause: unknown) {
    super(cause instanceof Error ? cause.message : String(cause));
    this.name = "LoaderError";

    Object.setPrototypeOf(this, LoaderError.prototype);
  }
}

export const load = (id: wee.AggregateId): T.Effect<Has<EventLoader>, LoaderError, wee.RecordedEvent[]> =>
  pipe(
    T.do,
    T.bind("loader", () => T.service(EventLoader)),
    T.bind("events", ({ loader }) =>
      T.tryCatchPromise(
        () => loader.load(id),
        (e) => new LoaderError(e)
      )
    ),
    T.map(({ events }) => events)
  );

export class PublisherError extends Error {
  constructor(public override readonly cause: unknown) {
    super(cause instanceof Error ? cause.message : String(cause));
    this.name = "PublisherError";

    Object.setPrototypeOf(this, PublisherError.prototype);
  }
}
export const publish = (
  aggregate: wee.AggregateId,
  events: wee.DomainEvent | wee.DomainEvent[],
  options?: wee.EventStore.PublishOptions
): T.Effect<
  Has<EventPublisher>,
  wee.ExpectedRevisionConflictError | wee.RevisionConflictError | PublisherError,
  wee.Revision
> =>
  pipe(
    T.do,
    T.bind("publisher", () => T.service(EventPublisher)),
    T.bind("revision", ({ publisher }) =>
      T.tryCatchPromise(
        async () => {
          return publisher.publish(aggregate, events, options);
        },
        (e) => new PublisherError(e)
      )
    ),
    T.map(({ revision }) => revision)
  );
