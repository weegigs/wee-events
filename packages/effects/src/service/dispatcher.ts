import * as wee from "@weegigs/events-core";

import * as z from "zod";

import * as T from "@effect-ts/core/Effect";
import { pipe } from "@effect-ts/core/Function";

type Payload = wee.Payload;

export type Command = wee.Payload;
export type State = Payload;

export type Handler<R, E, S extends Payload, C extends Command> = (
  entity: wee.Entity<S>,
  command: C
) => T.Effect<R, E, void>;

export type ValidationError = z.ZodError;

export class CommandValidationError extends Error {
  constructor(public readonly error: ValidationError) {
    super("command validation failed");
    this.name = "CommandValidationError";

    Object.setPrototypeOf(this, CommandValidationError.prototype);
  }
}

export class HandlerNotFound extends Error {
  constructor(public readonly path: string) {
    super(`Handler not found for path ${path}`);
    this.name = "HandlerNotFound";

    Object.setPrototypeOf(this, HandlerNotFound.prototype);
  }
}

const validate = <T>(schema: z.Schema<T>, value: T): T.IO<CommandValidationError, T> => {
  return T.suspend(() => {
    const result = schema.safeParse(value);
    if (!result.success) {
      return T.fail(new CommandValidationError(result.error));
    }

    return T.succeed(result.data);
  });
};

function validated<R, E, S extends Payload, C extends Payload>(
  schema: z.Schema<C>,
  handler: Handler<R, E, S, C>
): Handler<R, E | CommandValidationError, S, C> {
  return (entity: wee.Entity<S>, command: C) =>
    pipe(
      T.do,
      T.bind("command", () => validate(schema, command)),
      T.bind("result", () => handler(entity, command)),
      T.map(({ result }) => result)
    );
}

export type Target<T extends Payload = Payload> = wee.Entity<T> | wee.AggregateId;

export interface Dispatcher<R, E, State extends Payload> {
  dispatch(
    path: string,
    entity: Target<State>,
    command: Command
  ): T.Effect<R, E | HandlerNotFound | CommandValidationError, void>;

  handler: <R1, E1, C extends Command>(
    path: string,
    schema: z.Schema<C>,
    h: Handler<R1, E1, State, C>
  ) => Dispatcher<R & R1, E | E1, State>;

  commands: () => Record<string, z.Schema<Payload>>;
}

type Handlers<R, E, State extends Payload> = Record<
  string,
  { schema: z.Schema<unknown>; handler: Handler<R, E, State, Command> }
>;

export namespace Dispatcher {
  const $make = <R, E, State extends Payload>(handlers: Handlers<R, E, State> = {}): Dispatcher<R, E, State> => {
    const dispatch = (
      path: string,
      entity: wee.Entity<State>,
      command: Command
    ): T.Effect<R, E | HandlerNotFound | CommandValidationError, void> => {
      const handler = handlers[path];
      if (undefined == handler) {
        return T.fail(new HandlerNotFound(path));
      }

      return handler.handler(entity, command);
    };

    const handler = <R1, E1, C extends Command>(
      path: string,
      schema: z.Schema<C>,
      h: Handler<R1, E1, State, C>
    ): Dispatcher<R & R1, E | E1, State> => {
      const y = {
        [path]: { schema, handler: validated(schema, h) },
      };

      const x = {
        ...handlers,
        ...y,
      } as Handlers<R & R1, E | E1, State>;

      return $make(x);
    };

    const commands = () => {
      const result: Record<string, z.Schema> = {};

      for (const path in handlers) {
        result[path] = handlers[path].schema;
      }

      return result;
    };

    return {
      dispatch,
      commands,
      handler,
    };
  };

  export const create = <State extends Payload>(): Dispatcher<unknown, never, State> => $make({});
}
