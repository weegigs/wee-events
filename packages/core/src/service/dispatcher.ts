import * as z from "zod";

import { EventStore } from "../store";
import { Entity, Payload } from "../types";
import { State } from "./state";

export type Command = Payload;

export type Environment = Record<string, unknown>;

export type Publisher = { publish: EventStore.Publisher };

export type Handler<R extends Environment, S extends State, C extends Command> = (
  environment: R,
  entity: Entity<S>,
  command: C
) => Promise<void>;

export type ValidationError = z.ZodError;

export class CommandValidationError extends Error {
  constructor(public readonly error: ValidationError) {
    super("command validation failed");
    this.name = "CommandValidationError";

    Object.setPrototypeOf(this, CommandValidationError.prototype);
  }
}

export class HandlerNotFound extends Error {
  constructor(public readonly path: string) {
    super(`Handler not found for path ${path}`);
    this.name = "HandlerNotFound";

    Object.setPrototypeOf(this, HandlerNotFound.prototype);
  }
}

const validate = <T>(schema: z.Schema<T>, value: unknown): T => {
  const result = schema.safeParse(value);

  if (!result.success) {
    throw new CommandValidationError(result.error);
  }

  return result.data;
};

export interface Dispatcher<S extends State> {
  dispatch(path: string, entity: Entity<S>, command: Command): Promise<void>;
}

export namespace Dispatcher {
  export interface Description {
    commands: Record<string, z.Schema<Payload>>;
  }
}

export interface DispatcherDescription<R extends Environment, S extends State> {
  commands(): Record<string, z.Schema<Payload>>;
  dispatcher(environment: R): Dispatcher<S>;
}

export namespace DispatcherDescription {
  class Factory<R extends Environment, S extends State> {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    private handlers: Record<string, { schema: z.Schema<any>; handler: Handler<any, S, any> }> = {};

    constructor(
      path: string,
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      schema: z.Schema<any>,
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      handler: Handler<R, S, any>,
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      existing: Record<string, { schema: z.Schema<any>; handler: Handler<any, S, any> }> = {}
    ) {
      this.handlers = {
        ...existing,
        [path]: { schema, handler },
      };
    }

    handler: <R1 extends Environment, C extends Payload>(
      path: string,
      schema: z.ZodType<C, z.ZodTypeDef, C>,
      handler: Handler<R1, S, C>
    ) => Factory<R & R1, S> = <R1 extends Environment, C extends Payload>(
      path: string,
      schema: z.ZodType<C, z.ZodTypeDef, C>,
      handler: Handler<R1, S, C>
    ) => {
      return new Factory<R & R1, S>(path, schema, handler, this.handlers);
    };

    description: () => DispatcherDescription<R, S> = () => {
      return {
        dispatcher: (environment: R) => ({
          dispatch: async (path: string, entity: Entity<S>, command: Payload) => {
            const handler = this.handlers[path];

            if (!handler) {
              throw new HandlerNotFound(path);
            }

            await handler.handler(environment, entity, validate(handler.schema, command));
          },
        }),
        commands: () => Object.fromEntries(Object.entries(this.handlers).map(([path, { schema }]) => [path, schema])),
      };
    };
  }

  /**
   * Creates a dispatch factory that can be used to create a dispatcher
   *
   * @param path the path to connect the handler to
   * @param schema the zod schema used to validate the command
   * @param handler the handler used to execute the command
   * @returns a dispatch factory that can be used to create a dispatcher
   */
  export function handler<R extends Environment, S extends State, P extends string, C extends Command>(
    path: P,
    schema: z.Schema<C>,
    handler: Handler<R, S, C>
  ): Factory<R, S> {
    return new Factory(path, schema, handler, {});
  }
}

type HF<R extends Environment, P extends string, S extends State, C extends Command> = {
  path: P;
  schema: z.ZodType<C, z.ZodTypeDef, C>;
  handler: Handler<R, S, C>;
};

export function cf<R extends Environment, P extends string, S extends State, C extends Command>(
  hf: HF<R, P, S, C>
): (environment: R, path: P, state: Entity<S>, command: C) => Promise<void> {
  return async (environment: R, _path: P, state: Entity<S>, command: C) => {
    hf.handler(environment, state, validate(hf.schema, command));
  };
}
