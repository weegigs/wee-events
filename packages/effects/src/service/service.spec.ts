import _ from "lodash";

import * as T from "@effect-ts/core/Effect";
import { pipe } from "@effect-ts/core/Function";

import * as wee from "@weegigs/events-core";

import * as dispatcher from "./dispatcher";
// import * as loader from "./loader";
import * as store from "../event-store";

import * as receipts from "./sample/receipts";

const id = { key: "test", type: "receipt" };
const ms = new wee.MemoryStore();

describe("describe receipt service", () => {
  beforeEach(() => ms.clear());

  it("should load an empty entity", async () => {
    const program = pipe(receipts.service.load(id), T.provideService(store.EventLoader)(ms));

    const result = await T.runPromise(program);

    expect(result.state.total).toEqual(0);
    expect(result.revision).toEqual(wee.Revision.Initial);
    expect(result).toMatchSnapshot();
  });

  it.skip("should load an empty entity if the aggregate only contains unknown events", async () => {
    ms.publish(id, { type: "unknown", data: { value: "something" } });

    const program = pipe(receipts.service.load(id), T.provideService(store.EventLoader)(ms));
    const result = await T.runPromise(program);

    expect(result.state.total).toEqual(0);
    expect(result.revision).toEqual(wee.Revision.Initial);
    expect(result).toMatchSnapshot();
  });

  it("should load an entity with existing events", async () => {
    ms.publish(id, { type: "added", data: { amount: 10 } });

    const program = pipe(receipts.service.load(id), T.provideService(store.EventLoader)(ms));
    const result = await T.runPromise(program);

    expect(result.state.total).toEqual(10);
    expect(result.revision).not.toEqual(wee.Revision.Initial);
    expect(_.omit(result, "revision")).toMatchSnapshot();
  });

  it("should skip unknown event types", async () => {
    ms.publish(id, { type: "added", data: { amount: 10 } });
    ms.publish(id, { type: "unknown", data: { value: "something" } });

    const program = pipe(receipts.service.load(id), T.provideService(store.EventLoader)(ms));
    const result = await T.runPromise(program);

    expect(result.state.total).toEqual(10);
    expect(result.revision).not.toEqual(wee.Revision.Initial);
    expect(_.omit(result, "revision")).toMatchSnapshot();
  });

  it("should reject unknown paths", async () => {
    ms.publish(id, { type: "added", data: { amount: 10 } });

    const update = receipts.service.execute("add/it", id, { amount: 5 });

    const program = pipe(update, T.provideService(store.EventLoader)(ms), T.provideService(store.EventPublisher)(ms));
    expect(T.runPromise(program)).rejects.toBeInstanceOf(dispatcher.HandlerNotFound);
  });

  it("should validate command structure", async () => {
    ms.publish(id, { type: "added", data: { amount: 10 } });

    const update = receipts.service.execute("add", id, { value: 5 });

    const program = pipe(update, T.provideService(store.EventLoader)(ms), T.provideService(store.EventPublisher)(ms));
    expect(T.runPromise(program)).rejects.toBeInstanceOf(dispatcher.CommandValidationError);
  });

  it("should update an entity", async () => {
    ms.publish(id, { type: "added", data: { amount: 10 } });

    const update = receipts.service.execute("add", id, { amount: 5 });

    const program = pipe(update, T.provideService(store.EventLoader)(ms), T.provideService(store.EventPublisher)(ms));
    const result = await T.runPromise(program);

    expect(result.state.total).toEqual(15);
    expect(result.revision).not.toEqual(wee.Revision.Initial);
    expect(_.omit(result, "revision")).toMatchSnapshot();
  });

  it("should create an entity", async () => {
    const update = receipts.service.execute("add", id, { amount: 5 });

    const program = pipe(update, T.provideService(store.EventLoader)(ms), T.provideService(store.EventPublisher)(ms));
    const result = await T.runPromise(program);

    expect(result.state.total).toEqual(5);
    expect(result.revision).not.toEqual(wee.Revision.Initial);
    expect(_.omit(result, "revision")).toMatchSnapshot();
  });

  it("should propagate handler errors", async () => {
    ms.publish(id, { type: "added", data: { amount: 10 } });

    const update = receipts.service.execute("deduct", id, { amount: 15 });
    const program = pipe(update, T.provideService(store.EventLoader)(ms), T.provideService(store.EventPublisher)(ms));

    expect(T.runPromise(program)).rejects.toBeInstanceOf(receipts.InsufficientBalanceError);
  });
});
