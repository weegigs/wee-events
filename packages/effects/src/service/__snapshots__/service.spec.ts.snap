// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`describe receipt service should create an entity 1`] = `
Object {
  "aggregate": Object {
    "key": "test",
    "type": "receipt",
  },
  "state": Object {
    "id": "receipt.test",
    "total": 5,
  },
  "type": "receipt",
}
`;

exports[`describe receipt service should load an empty entity 1`] = `
Object {
  "aggregate": Object {
    "key": "test",
    "type": "receipt",
  },
  "revision": "00000000000000000000000000",
  "state": Object {
    "id": "receipt.test",
    "total": 0,
  },
  "type": "receipt",
}
`;

exports[`describe receipt service should load an entity with existing events 1`] = `
Object {
  "aggregate": Object {
    "key": "test",
    "type": "receipt",
  },
  "state": Object {
    "id": "receipt.test",
    "total": 10,
  },
  "type": "receipt",
}
`;

exports[`describe receipt service should update an entity 1`] = `
Object {
  "aggregate": Object {
    "key": "test",
    "type": "receipt",
  },
  "state": Object {
    "id": "receipt.test",
    "total": 15,
  },
  "type": "receipt",
}
`;
